src/sage/topology/moment_angle_complex.py

```diff
@@ -814,3 +819,712 @@ def has_trivial_lowest_deg_massey_product(self):
+            sage: b.cup_product(a)
+            h^{6,0}
+            sage: a.cup_product(b)
+            -h^{6,0}
+
+            sage: Z = MomentAngleComplex([[1,2], [1,4], [2,3],
+            ....:                         [3,5], [5,6], [4,5], [1,6]])
+            sage: H = Z.cohomology_ring(GF(2))
+            sage: one = H.one()
+            sage: a = H.basis()[4, 3]
+            sage: b = H.basis()[4, 6]
+            sage: a.cup_product(b)
+            h^{8,1}
+            sage: b.cup_product(a)
+            h^{8,1}
+            sage: one.cup_product(a) == a.cup_product(one)
+            True
+        """
+        # First we check whether either one of the elements
+        # represents a unit of this cohomology ring;
+        # This is necessary because the multiplicative identity
+        # is the only element whose corresponding cochain has degree -1
+        if li == (0, 0):
+            return self.basis()[ri]
+        elif ri == (0, 0):
+            return self.basis()[li]
+
+        from sage.topology.simplicial_complex import Simplex
+        from sage.homology.chain_homotopy import ChainContraction
+
+        left = self._gens[li[0]][li[1]]
+        right = self._gens[ri[0]][ri[1]]
+
+        # Sets of vertices of subcomplexes from which
+        # these cohomology classes originate
+        set_left = left[0]
+        set_right = right[0]
+        # Because of the definition of multiplication
+        if not set_left.isdisjoint(set_right):
+            return self.zero()
+
+        # We extract the cocycle monomials in
+        # order to loop over them
+        left_cocycles = left[2][1].monomials()
+        right_cocycles = right[2][1].monomials()
+        res = self.zero()
+        for left_cocycle in left_cocycles:
+            for right_cocycle in right_cocycles:
+                # We acquire the set of vertices of cocyles,
+                # because of the union
+                left_cocycle_vertices = left_cocycle.leading_support().set()
+                right_cocycle_vertices = right_cocycle.leading_support().set()
+                union = set_left.union(set_right)
+                res_union = left_cocycle_vertices.union(right_cocycle_vertices)
+                subcomplex_union = self._complex._simplicial_complex.generated_subcomplex(union, is_mutable=False)
+
+                # Because we special-cased the multiplicative identity
+                # at the beginning of the method, we know that both
+                # left[1] and right[1] are at least 0 (they represent
+                # the degree of the corresponding cochains). Therefore,
+                # deg is at least 1.
+                deg = left[1] + right[1] + 1
+                cochains_basis = subcomplex_union.n_chains(deg, cochains=True).basis()
+                # If the resulting cochain exists, then we
+                # add it to the result; otherwise, we add 0
+                if cochains_basis.has_key(Simplex(res_union)):
```

Comment:
```suggestion
                X = Simplex(res_union)
                if X in cochains_basis:
```
since `cochains_basis` is a `dict`, right? This way you don't have to recreate the `Simplex` below either. Although I would do this as
```
if X not in cochains_basis:
    continue
[rest of code]
```